community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/InMemoryCountsStore.java

```diff
@@ -62,6 +67,53 @@ public InMemoryCountsStore()
     }
 
     @Override
+    public void replace( CountsKey key, long[] replacement )
+    {
+        lock.readLock().lock();
+        try
+        {
+            if ( snapshot != null )
+            {
+                throw new IllegalStateException(
+                        "Cannot alter count store outside of a transaction while a snapshot is processing." );
+            }
+            map.put( key, replacement );
+        }
+        finally
+        {
+            lock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public void update( CountsKey key, long[] delta )
+    {
+        lock.readLock().lock();
+        try
+        {
+            if ( snapshot != null )
+            {
+                throw new IllegalStateException(
+                        "Cannot alter count store outside of a transaction while a snapshot is processing." );
+            }
+            map.compute( key, ( k, v ) -> {
+                if ( v == null )
+                {
+                    return Arrays.copyOf( delta, delta.length );
```

Comment:
do we really need to copy always incoming array? that will create tons of array copies, not only here but on all places where we copy and apply updates?
Maybe final data structure to hold values?